File: src/main/kotlin/io/energyconsumptionoptimizer/mapservice/domain/ports/SmartFurnitureHookupService.kt
package io.energyconsumptionoptimizer.mapservice.domain.ports

import io.energyconsumptionoptimizer.mapservice.domain.SmartFurnitureHookup
import io.energyconsumptionoptimizer.mapservice.domain.SmartFurnitureHookupID
import io.energyconsumptionoptimizer.mapservice.domain.ZoneID

/**
 * Service interface for managing smart furniture hookup operations.
 */

interface SmartFurnitureHookupService {
    /**
     * Creates a new smart furniture hookup at a specified position.
     *
     * Note: The zoneID should be assigned based on the position if is not provided.
     *
     * @param id The unique identifier for the smart furniture hookup
     * @param position The (x, y) coordinates of the hookup on the floor plan
     * @param zoneID The unique identifier of the zone containing this hookup, or null if not assigned to a zone
     * @return The created smart furniture hookup entity
     *
     * @throws io.energyconsumptionoptimizer.mapservice.domain.errors.SmartFurnitureHookupAlreadyExistsException
     */
    suspend fun createSmartFurnitureHookup(
        id: SmartFurnitureHookupID,
        position: Pair<Double, Double>,
        zoneID: ZoneID?,
    ): SmartFurnitureHookup

    /**
     * Retrieves all smart furniture hookups across the floor plan.
     *
     * @return A list of all smart furniture hookups, or an empty list if none exist
     */
    suspend fun getSmartFurnitureHookups(): List<SmartFurnitureHookup>

    /**
     * Retrieves a specific smart furniture hookup by its unique identifier.
     *
     * @param id The unique identifier of the smart furniture hookup
     * @return The smart furniture hookup if found, null otherwise
     */
    suspend fun getSmartFurnitureHookup(id: SmartFurnitureHookupID): SmartFurnitureHookup?

    /**
     * Updates an existing smart furniture hookup's position and/or zone assignment.
     *
     * Only non-null parameters will be updated. If a parameter is null, the existing value is retained.
     *
     * Note: The zoneID should be reassigned if the position changed and a new zoneId is not provided.
     *
     * @param id The unique identifier of the smart furniture hookup to update
     * @param position The new (x, y) coordinates, or null to keep the existing position
     * @param zoneID The new zone identifier, or null to keep the existing zone assignment
     * @return The updated smart furniture hookup entity
     *
     * @throws io.energyconsumptionoptimizer.mapservice.domain.errors.SmartFurnitureHookupIDNotFoundException
     */
    suspend fun updateSmartFurnitureHookup(
        id: SmartFurnitureHookupID,
        position: Pair<Double, Double>?,
        zoneID: ZoneID?,
    ): SmartFurnitureHookup

    /**
     * Deletes a smart furniture hookup.
     *
     * @param id The unique identifier of the smart furniture hookup to delete
     * @throws NoSuchElementException if the smart furniture hookup does not exist
     *
     * @throws io.energyconsumptionoptimizer.mapservice.domain.errors.SmartFurnitureHookupIDNotFoundException
     */
    suspend fun deleteSmartFurnitureHookup(id: SmartFurnitureHookupID)
}

File: src/application/inbound/SmartFurnitureHookupService.ts
import type { SmartFurnitureHookup } from "@domain/entities/SmartFurnitureHookup";

/**
 * @inbound
 * Service interface for managing smart furniture hookup placement operations.
 *
 * Note: The "smart furniture hookup" in this context is a positioned device on
 * the floor plan map; its ID originates from the separate smart-furniture-hookup
 * service.
 */
export interface SmartFurnitureHookupService {
  /**
   * Places a smart furniture hookup on the floor plan at the given position.
   *
   * If `zoneID` is not provided the zone is auto-detected from the hookup's
   * coordinates. If `zoneID` is provided, the referenced zone must exist.
   *
   * @returns The saved SmartFurnitureHookup, or an Error (IdEmptyError,
   *          InvalidCoordinateError, ZoneNotFoundError,
   *          SmartFurnitureHookupAlreadyExistsError).
   */
  createSmartFurnitureHookup(id: string): Promise<SmartFurnitureHookup | Error>;

  /**
   * Retrieves all smart furniture hookups on the floor plan.
   */
  getSmartFurnitureHookups(): Promise<SmartFurnitureHookup[]>;

  /**
   * Retrieves a single smart furniture hookup by its unique identifier.
   *
   * @returns The SmartFurnitureHookup if found, `null` otherwise.
   */
  getSmartFurnitureHookup(id: string): Promise<SmartFurnitureHookup | Error>;

  /**
   * Updates the position and/or zone assignment of a smart furniture hookup.
   *
   * Only supplied (non-undefined) parameters are updated. When position changes
   * and no explicit `zoneID` is given, the zone is recalculated automatically.
   * Pass `zoneID: null` to explicitly unassign the hookup from its zone.
   *
   * @returns The updated SmartFurnitureHookup, or an Error (IdEmptyError,
   *          InvalidCoordinateError, SmartFurnitureHookupNotFoundError,
   *          ZoneNotFoundError).
   */
  updateSmartFurnitureHookup(
    id: string,
    position?: [number, number],
    zoneID?: string,
  ): Promise<SmartFurnitureHookup | Error>;

  /**
   * Removes a smart furniture hookup from the floor plan.
   *
   * @returns `undefined` on success, or an Error (IdEmptyError,
   *          SmartFurnitureHookupNotFoundError).
   */
  deleteSmartFurnitureHookup(id: string): Promise<undefined | Error>;
}
